### PR TITLE
Apply target regions consistently in all SolidMechanicsLagrangianFEM kernels

### LINKED ISSUE #1140: [Feature] Clarification needed for using `SolidMechanicsLagrangianFEM` with multiple target regions
I am trying to run a single-phase poroelastic simulation for @AntoineMazuyer 's mesh. Antoine included multiple regions (Reservoir, Underburden, Overburden, Sea) and would like to simulate flow in Reservoir and mechanics in Reservoir, Underburden, and Overburden. I am struggling with the treatment of Sea, where we don't want to simulate anything (I removed the mesh image below because I am not sure I am allowed to use it, but I can send it on Slack)

Here is what I tried in this [file](https://totalcse.box.com/s/npxphmvwl23gz9ed1roikega3bso4noh):
  
1. If I don't define any `ElementRegion` in Sea, I get negative `localRows` in the computation of the residual norm for the mechanical problem [here](https://github.com/GEOSX/GEOSX/blob/92977f0fddf30be497e88777b25836ee9113d0af/src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp#L1115). If I guard this against `localRows` that are between 0 and `localRhs.size()` everything works well.

 2.  If I define an `ElementRegion` in Sea, then if I understand correctly, the `dofManager` defines positive `dofNumbers` in Sea, so I assume that `localRhs` and `localMatrix` will include some entries for the Sea region (which is not what I want).

@klevzoff has good suggestions of modifications that would make this case work for options 1. and 2. Thanks @klevzoff for your help.

@rrsettgast @klevzoff Should we make it clear that users have to provide a mesh in which the `solidMechanicsLagrangianFEM` solver will be applied everywhere? Or should we support the use case in which a region is not the `targetRegion` of any of the solvers?

### PR DESCRIPTION
**Pull request by francoishamon**
The PR makes two modifications in `SolidMechanicsLagrangianFEM` to make sure that target regions are applied consistently in all kernels:
- In `SetupDofs`, the physics solver now passes the target regions to the `DofManager`.
- In `CalculateResidualNorm`, the norm is computed over the target nodes formed by the union of `m_nonSendOrReceiveNodes` and `m_sendOrReceiveNodes`.

One thing to note is that for some coupled solvers (namely, `LagrangianContactSolver` and `HydrofractureSolver`) this change requires specifying `targetRegions="{ Region, Fracture }"` instead of just `targetRegions="{ Region }"` in the `SolidMechanicsLagrangianFEM` block of the xml file. See the discussion in issue #1140 about this.

I modified the integrated test `gravity.xml` to have two regions, and apply `SolidMechanicsLagrangianFEM` on only one of them. This is a little bit un-natural, but this is the situation that motivated the issue. More realistic use cases will come when we can couple mechanics with wells, for example.

Related to https://github.com/GEOSX/integratedTests/pull/96.

Resolves #1140.

### CODE REVIEW

**Review by rrsettgast — APPROVED** [review]

File: src/coreComponents/physicsSolvers/solidMechanics/integratedTests/gravity.xml [hunk 1]
```diff
@@ -4,14 +4,14 @@
   <Mesh>
     <InternalMesh
       name="mesh1"
-      cellBlockNames="{ cb1 }"
+      cellBlockNames="{ cb1, cb2 }"
       elementTypes="{ C3D8 }"
       nx="{ 1 }"
       ny="{ 1 }"
-      nz="{ 100 }"
+      nz="{ 100, 100 }"
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/solidMechanics/integratedTests/gravity.xml]: Done, the problem now has 20 elements.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 2]
```diff
@@ -1110,12 +1115,15 @@ SolidMechanicsLagrangianFEM::
...
 
-  forAll< parallelDevicePolicy<> >( nodeManager.size(),
-                                    [localRhs, localSum, dofNumber, rankOffset, ghostRank] GEOSX_HOST_DEVICE ( localIndex const k )
+  SortedArrayView< localIndex const > const & targetNodes = m_targetNodes.toViewConst();
+
+  forAll< parallelDevicePolicy<> >( targetNodes.size(),
+                                    [localRhs, localSum, dofNumber, rankOffset, ghostRank, targetNodes] GEOSX_HOST_DEVICE ( localIndex const k )
   {
-    if( ghostRank[k] < 0 )
+    if( ghostRank[targetNodes[k]] < 0 )
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp]: Done

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 3]
```diff
@@ -1110,12 +1115,15 @@ SolidMechanicsLagrangianFEM::
...
     {
-      localIndex const localRow = LvArray::integerConversion< localIndex >( dofNumber[k] - rankOffset );
+      localIndex const localRow = LvArray::integerConversion< localIndex >( dofNumber[targetNodes[k]] - rankOffset );
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp]: Done

**Comment by rrsettgast** [comment]
@klevzoff Can you review and confirm you are good with the changes?

**Review by klevzoff — APPROVED** [review]
Nice!

My only concern is the cost of constructing `m_targetNodes`, which seems to be O(N^2) worst case. Same goes for to `m_sendOrReceiveNodes` and `m_nonSendOrReceiveNodes` to be fair, but I'm assuming you guys are ok with those @rrsettgast @corbett5 . `m_targetNodes` could be instead constructed by merging those two sorted arrays in O(N), but that might need to be implemented as an `LvArray` method.

**Comment by corbett5** [comment]
I think an easy and efficient way is that `m_sendOrReceiveNodes` and `m_nonSendOrReceiveNodes` are constructed using a `std::set` then you can just call `m_sendOrReceiveNodes.insert(set.begin(), set.end())` and that should be efficient especially if `m_sendOrReceiveNodes` was previously empty.

As for constructing `m_targetNodes` there are a few ways to go. One would be to just copy `m_nonSendOrReceiveNodes` and then insert from `m_sendOrReceiveNodes` which would be `O(N)`. The best way would probably be to use `std::set::merge` which is only in c++17 :( then create `m_targetNodes` from that.

https://en.cppreference.com/w/cpp/container/set/merge

There's also `std::set_union` but that would involve either creating a new container to hold the merged sets (could be a std::vector) or adding some insert iterator capability to `SortedArray`.

**Comment by rrsettgast** [comment]
@klevzoff @corbett5 I missed the issue with filling of `m_targetNodes`. I think the intent of `m_targetNodes` is to contain all the nodes that have a DOF assigned to them? If that is the case, the `DofManager` already has done the work. so we could do something like :
```
string const dofKey = dofManager.getKey( keys::TotalDisplacement );
arrayView1d< globalIndex const > const & dofNumber = nodeManager.getReference< globalIndex_array >( dofKey );
arrayView1d< integer const > const & ghostRank = nodeManager.ghostRank();
m_targetNodes.resize(dofNumber.size());
localIndex counter = 0;
forAll< serialPolicy >( dofNumber.size(), [&] ( localIndex const a )
{
  if( dofNumber[a] >= 0 && ghostRank[a] <=-1 ) // I think you only want locally owned nodes that are assigned a dog?
  {
    m_targetNodes[counter] = a;
    ++counter;
  }
} );
m_targetNodes.resize(counter);
```

**Comment by klevzoff** [comment]
@rrsettgast yes, that's a good observation. 

By the way, `m_targetNodes` should in theory also be used in all other node-based kernel launches (e.g. velocity updates in explicit Newmark step), but this PR only addresses implicit solve. There's a trade-off to be made between increased memory bandwidth for those kernels and reducing the size of the launch in cases where only part of the mesh contains active dofs (which is probably quite rare).

**Comment by rrsettgast** [comment]
@klevzoff That is a good point. I had thought we were using `m_sendOrReceiveNodes` and `m_nonSendOrReceiveNodes` as the filter for these updates, but they are only used on the update of velocity from mid-step to the end-of-step.

**Comment by rrsettgast** [comment]
@francoishamon Ready to merge? Can you run tests on Lassen?

**Comment by francoishamon** [comment]
@rrsettgast At some point a couple of weeks ago I had implemented what Ben suggested above to accelerate set insertion, but never committed the modifications because the discussion on the PR was still going on. I am going to commit this modification, update the integratedTests, and test on Lassen. I will ping you when it is ready.

**Comment by francoishamon** [comment]
@rrsettgast I made this PR ready to be merged after the integratedTests repository is updated (I saw that @castelletto1 's PR is also waiting so I don't mind rebaselining again if Nicola merges first).

On Lassen I get the following failures:
```FAIL RUN : 6                                                                                                                                  
 (  KGD_ZeroViscosity_06 Sneddon_01 PhaseFieldFracture_DamageOnly_01 PhaseFieldFracture_DamageOnly_08 PhaseFieldFracture_DamageAndLoad_01 PhaseFieldFracture_DamageAndLoad_08  )
```
These tests seem to be failing on `develop` in the nightly tests as well.

**Pull request merged**